### PR TITLE
Apply `single_version_override` patches to module files

### PR DESCRIPTION
**Pull request by fmeum**
Work towards #19301

RELNOTES: Patches to the module file in `single_version_override` are now effective as long as the patch file lies in the root module.

### CODE REVIEW

File: src/test/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunctionTest.java [hunk 1]
```diff
@@ -1686,4 +1690,89 @@ public void testInvalidUseExtensionLabel() throws Exception {
...
+        +++ b/MODULE.bazel
+        @@ -1,1 +1,3 @@
+         module(name='bbb',version='1.0',bazel_compatibility=[">=7.0.0"])
+        +bazel_dep(name='ccc',version='2.0')
+        """);
+    differencer.invalidate(
+        ImmutableList.of(
+            FileValue.key(RootedPath.toRootedPath(Root.fromPath(rootDirectory), otherPatch))));
+
+    result = evaluator.evaluate(ImmutableList.of(moduleFileKey), evaluationContext);
```
Comment (fmeum) [inline src/test/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunctionTest.java:1767]: @Wyverald This call returns the stale value and I don't understand why. I can tell that the `difference.invalidate` call actually marks the correct nodes for invalidation, but somehow `ModuleFileFunction` isn't rerun even though it has a dep registered.

Do you have any idea how I could debug this further?

Comment (fmeum) [reply src/test/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunctionTest.java:1768]: Nvm, fixed

File: src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java [hunk 2]
```diff
@@ -417,10 +417,25 @@ private static void checkFilesStatusForPatching(
...
+   * Apply a patch file under a directory
+   *
+   * @param patchFile the patch file to apply
+   * @param strip the number of leading components to strip from file path in the patch file
+   * @param outputDirectory the directory to apply the patch file to
+   * @param singleFile if not null, only apply the parts of the patch file that apply to this file.
+   *     Renaming the file is not supported in this case.
+   */
+  public static void apply(
+      Path patchFile, int strip, Path outputDirectory, @Nullable Path singleFile)
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java:437]: this new optional parameter changes the method's behavior quite significantly. Could we rename this method to `applyToSingleFile` and make the `@Nullable` version private? (and maybe rename `singleFile` the parameter to `fileToPatch` or something)

**Review by Wyverald — APPROVED** [review]
very nice! using an in-memory file system is clever. I just have some stylistic comments.

**Review by Wyverald — APPROVED** [review]

**Comment by fmeum** [comment]
@bazel-io fork 7.4.0

**Comment by iancha1992** [comment]
@fmeum can you please squash the commits since it has a third party change? Thanks

@bazelbuild/triage

**Comment by fmeum** [comment]
I fixed the merge conflict.

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!

**Pull request closed without merging**